### PR TITLE
genesis: add support for arbitrary initial height

### LINKED ISSUE #2543: Allow genesis file to start from non-zero height w/ prev block header
We should introduce the ability for a genesis file to start from a different block height. This is to create a sense of continuity when governance is approving hard forks. It also allows those who were unaware of the upgrade to not be confused by the sudden move back to block zero. (We want to encourage protocol changes to happen, so this sort of social confusion would be a blocker on such a change)

If the block number is non-zero, then the chain should also include the header of the previous block, just to ensure that this new genesis does tie into the original genesis file. The previous block has to have 0 txs (or w/e txs were there just don't get ran), because they may be invalidated under this upgrade.

/cc @ebuchman

### PR DESCRIPTION
**Pull request by erikgrinaker**
Adds a genesis parameter `initial_height` which specifies the initial block height, as well as ABCI `RequestInitChain.InitialHeight` to pass it to the ABCI application, and `State.InitialHeight` to keep track of the initial height throughout the code. Fixes #2543, based on [RFC-002](https://github.com/tendermint/spec/pull/119). Spec changes in https://github.com/tendermint/spec/pull/135.

### CODE REVIEW

**Comment by marbar3778** [comment]
nit: the returned error from rpc for a nonexisting block is: `"height 500 is not available, blocks pruned at height 500000"` but the blocks weren't pruned, the current blockchain started at a different height. Not sure if the error should be returned with a more general sentence or if a different error could be thrown?

**Comment by erikgrinaker** [comment]
Good catch, I've reworded the error message as `height %v is not available, lowest height is %v`

File: cmd/tendermint/commands/testnet.go [hunk 1]
```diff
@@ -175,6 +178,7 @@ func testnetFiles(cmd *cobra.Command, args []string) error {
 		ChainID:         "chain-" + tmrand.Str(6),
 		ConsensusParams: types.DefaultConsensusParams(),
 		GenesisTime:     tmtime.Now(),
+		InitialHeight:   initialHeight,
```
Comment (cmwaters) [inline cmd/tendermint/commands/testnet.go:181]: Do you want to add the InitialHeight argument to the cmd/init.go file for initializing individual nodes

File: cmd/tendermint/commands/testnet.go [hunk 2]
```diff
@@ -50,6 +51,8 @@ func init() {
 		"Directory to store initialization data for the testnet")
 	TestnetFilesCmd.Flags().StringVar(&nodeDirPrefix, "node-dir-prefix", "node",
 		"Prefix the directory name for each node with (node results in node0, node1, ...)")
+	TestnetFilesCmd.Flags().Int64Var(&initialHeight, "initial-height", 0,
```
Comment (cmwaters) [inline cmd/tendermint/commands/testnet.go:54]: Do we also want a flag added for the initializing an individual node (init.go)

File: proto/tendermint/state/types.proto [hunk 3]
```diff
@@ -41,7 +41,8 @@ message State {
   Version version = 1 [(gogoproto.nullable) = false];
 
   // immutable
-  string chain_id = 2 [(gogoproto.customname) = "ChainID"];
+  string chain_id       = 2 [(gogoproto.customname) = "ChainID"];
+  int64  initial_height = 14;
```
Comment (marbar3778) [inline proto/tendermint/state/types.proto:45]: is this aimed at grouping releative fields?

Comment (erikgrinaker) [reply proto/tendermint/state/types.proto:45]: Yes, these are all immutable.

File: cmd/tendermint/commands/testnet.go (hunk 2 above)
Comment (erikgrinaker) [reply cmd/tendermint/commands/testnet.go:54]: `tendermint init` doesn't appear to take any options at all (only global options), so I guess not? I suppose people configure the node by editing the genesis file?

Not to say we shouldn't add options, but in that case we should also add options for e.g. the chain ID and other settings. Not sure if this is needed.

Comment (cmwaters) [reply cmd/tendermint/commands/testnet.go:54]: Yeah I guess leave it as is

File: consensus/state.go [hunk 4]
```diff
@@ -555,27 +555,33 @@ func (cs *State) updateToState(state sm.State) {
...
-		return
+	if !cs.state.IsEmpty() {
+		if cs.state.LastBlockHeight > 0 && cs.state.LastBlockHeight+1 != cs.Height {
+			// This might happen when someone else is mutating cs.state.
+			// Someone forgot to pass in state.Copy() somewhere?!
+			panic(fmt.Sprintf("Inconsistent cs.state.LastBlockHeight+1 %v vs cs.Height %v",
+				cs.state.LastBlockHeight+1, cs.Height))
+		}
+		if cs.state.LastBlockHeight > 0 && cs.Height == cs.state.InitialHeight {
+			panic(fmt.Sprintf("Inconsistent cs.state.LastBlockHeight %v, expected 0 for initial height %v",
```
Comment (melekes) [inline consensus/state.go:566]: should we also output cs.Height here since we're using it above?

File: consensus/state.go [hunk 5]
```diff
@@ -595,13 +601,16 @@ func (cs *State) updateToState(state sm.State) {
...
 		// NOTE: when Tendermint starts, it has no votes. reconstructLastCommit
 		// must be called to reconstruct LastCommit from SeenCommit.
-		panic(fmt.Sprintf("LastCommit cannot be empty in heights > 1 (H:%d)",
+		panic(fmt.Sprintf("LastCommit cannot be empty after initial block (H:%d)",
 			state.LastBlockHeight+1,
 		))
 	}
 
 	// Next desired block height
 	height := state.LastBlockHeight + 1
```
Comment (melekes) [inline consensus/state.go:610]: I think this should be a func

File: state/store.go [hunk 6]
```diff
@@ -130,11 +130,16 @@ func saveState(db dbm.DB, state State, key []byte) {
 
 // BootstrapState saves a new state, used e.g. by state sync when starting from non-zero height.
 func BootstrapState(db dbm.DB, state State) error {
-	height := state.LastBlockHeight
-	saveValidatorsInfo(db, height, height, state.LastValidators)
-	saveValidatorsInfo(db, height+1, height+1, state.Validators)
-	saveValidatorsInfo(db, height+2, height+2, state.NextValidators)
-	saveConsensusParamsInfo(db, height+1, height+1, state.ConsensusParams)
+	height := state.LastBlockHeight + 1
```
Comment (melekes) [inline state/store.go:133]: to avoid duplication

File: consensus/reactor.go [hunk 7]
```diff
@@ -249,6 +249,14 @@ func (conR *Reactor) Receive(chID byte, src p2p.Peer, msgBytes []byte) {
 	case StateChannel:
 		switch msg := msg.(type) {
 		case *NewRoundStepMessage:
+			conR.conS.mtx.Lock()
+			initialHeight := conR.conS.state.InitialHeight
```
Comment (ebuchman) [inline consensus/reactor.go:253]: Isn't this immutable? Maybe we can load it once upfront and then not need to worry about the mutex?

Comment (erikgrinaker) [reply consensus/reactor.go:253]: Yeah, but I figured that'd be premature optimization unless we could demonstrate that there was actual mutex contention -- complexity (i.e. caching) is usually far more likely to cause problems. Although I suppose storing this during `consensus.NewState()` would pretty much always do the right thing.

Opened #6012, we can have a look after the refactor.

Comment (alexanderbez) [reply consensus/reactor.go:253]: Note, this was changed to a read-lock in the consensus reactor p2p refactor

**Pull request merged**